### PR TITLE
feat: implement pg_catalog upon local execution mode

### PR DESCRIPTION
**Pull request by yezizp2012**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

As title, implemented `pg_type`/`pg_namespace`/`pg_cast` and some other rules for ddl:
1. introduce a new `SysRowSeqScanExecutor` to support scanning on system table, only works in local execution mode.
2. init `pg_catalog` schema and system tables for all new created database, and drop them when database dropped.
3. refuse to create schema with prefix `pg_`.
4. refuse to create any relation under `pg_catalog`.

todo:
1. auto enable local mode when query on pg_catalog.
6. support more system tables.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
Part of #2954

### CODE REVIEW

File: src/frontend/src/binder/relation/table_or_source.rs [hunk 1]
```diff
@@ -58,25 +66,28 @@ impl Binder {
...
+                    };
+                    (
+                        Relation::SystemTable(Box::new(table)),
+                        sys_table_catalog.columns.clone(),
+                    )
+                } else {
+                    return Err(RwError::from(CatalogError::NotFound(
+                        "table",
+                        table_name.to_string(),
+                    )));
```
[the remark this answers is not in the document]
Comment (yezizp2012) [reply src/frontend/src/binder/relation/table_or_source.rs:88]: Good point.

**Review by lmatz — COMMENTED** [review]
Scheduler and Batch execution parts LGTM!

File: src/frontend/src/scheduler/local.rs [hunk 2]
```diff
@@ -42,15 +42,26 @@ pub struct LocalQueryExecution {
     query: Query,
     front_env: FrontendEnv,
     epoch: Option<u64>,
+
+    current_database: String,
+    current_user: String,
```
Comment (liurenjie1024) [inline src/frontend/src/scheduler/local.rs:47]: As a better practice, I would suggest to put these two fields into anther construct in `FrontendEnv`, for example, `AuthContext`

File: src/frontend/src/scheduler/task_context.rs [hunk 3]
```diff
@@ -16,21 +16,30 @@ use std::sync::Arc;
...
+use crate::catalog::pg_catalog::SysCatalogReaderImpl;
 use crate::session::FrontendEnv;
 
 /// Batch task execution context in frontend.
 #[derive(Clone)]
 pub struct FrontendBatchTaskContext {
     env: FrontendEnv,
+
+    current_database: String,
+    current_user: String,
```
Comment (liurenjie1024) [inline src/frontend/src/scheduler/task_context.rs:33]: As above comment

**Comment by liurenjie1024** [comment]
Generally LGTM

File: src/frontend/src/scheduler/local.rs (hunk 2 above)
Comment (yezizp2012) [reply src/frontend/src/scheduler/local.rs]: I introduced an `AuthContext` in `SessionImp` and passing it around. Put it into `FrontendEnv` is not quit reasonable, cuz all fields inside it are stateless.

**Review by xiangjinwu — APPROVED** [review]

**Comment by hengm3467** [comment]
@yezizp2012 @xiangjinwu Are these changes visible to users? Does it mean that in local execution mode, RisingWave supports casts, namespaces, and types exactly the same as in PG? Thanks.

**Comment by xiangjinwu** [comment]
> @yezizp2012 @xiangjinwu Are these changes visible to users? Does it mean that in local execution mode, RisingWave supports casts, namespaces, and types exactly the same as in PG? Thanks.

* These changes are visible to users.
* However, we do not want to encourage users to use them. They are here to support bigger user-friendly features that are not complete yet. PostgreSQL also exposes these "internal" things to user, but practically it adds little value to normal users.
* They are not exactly the same as in PG.

**Comment by hengm3467** [comment]
@xiangjinwu  Thanks for the detailed explanation! I think it makes senses that we do not disclose these details until the bigger features are completed.

**Pull request merged**